services/src/main/java/org/keycloak/broker/saml/SAMLEndpoint.java

```diff
@@ -371,13 +371,12 @@ protected Response handleLoginResponse(String samlResponse, SAMLDocumentHolder h
                     assertionElement = DocumentUtil.getElement(holder.getSamlDocument(), new QName(JBossSAMLConstants.ASSERTION.get()));
                 }
 
-                if (config.isWantAssertionsSigned() && config.isValidateSignature()) {
-                    if (!AssertionUtil.isSignatureValid(assertionElement, getIDPKeyLocator())) {
-                        logger.error("validation failed");
-                        event.event(EventType.IDENTITY_PROVIDER_RESPONSE);
-                        event.error(Errors.INVALID_SIGNATURE);
-                        return ErrorPage.error(session, Messages.INVALID_REQUESTER);
-                    }
+                if ((config.isWantAssertionsSigned() && !XMLSignatureUtil.isSigned(assertionElement.getOwnerDocument()))
```

Comment:
WantAssertionsSigned requires signed assertions while what is checked for validity in isSigned is the whole document (arbitrary element signature would suffice to satisfy the condition in that method). This has to be fixed to reflect WantAssertionsSigned logic.